rust/opendp-ffi/src/util.rs

```diff
@@ -1,61 +1,160 @@
+/// Builds a [`Type`] from a compact invocation, choosing an appropriate [`TypeContents`].
+/// * `t!(Foo)` => `TypeContents::PLAIN`
+/// * `t!((Foo, Bar))` => `TypeContents::TUPLE`
+/// * `t!([Foo; 10])` => `TypeContents::ARRAY`
+/// * `t!([Foo])` => `TypeContents::SLICE`
+/// * `t!(Foo<Bar>)` => `TypeContents::GENERIC`
+/// * `t!(Vec<primitive>)` => `TypeContents::VEC`
+macro_rules! t {
+    (Vec<$arg:ty>) => {
+        Type::new(
+            TypeId::of::<Vec<$arg>>(),
+            concat!("Vec<", stringify!($arg), ">"),
+            TypeContents::VEC(TypeId::of::<$arg>())
+        )
+    };
+    ($name:ident<$($args:ty),+>) => {
+        Type::new(
+            TypeId::of::<$name<$($args),+>>(),
+            concat!(stringify!($name), "<", stringify!($($args),+), ">"),
+            TypeContents::GENERIC { name: stringify!($name), args: vec![$(TypeId::of::<$args>()),+]}
+        )
+    };
+    ([$element:ty]) => {
+        Type::new(
+            TypeId::of::<[$element]>(),
+            concat!("[", stringify!($element), "]"),
+            TypeContents::SLICE(TypeId::of::<$element>())
+        )
+    };
+    ([$element:ty; $len:expr]) => {
+        Type::new(
+            TypeId::of::<[$element;$len]>(),
+            concat!("[", stringify!($element), "; ", stringify!($len), "]"),
+            TypeContents::ARRAY { element_id: TypeId::of::<$element>(), len: $len }
+        )
+    };
+    (($($elements:ty),+)) => {
+        Type::new(
+            TypeId::of::<($($elements),+)>(),
+            concat!("(", stringify!($($elements),+), ")"),
+            TypeContents::TUPLE(vec![$(TypeId::of::<$elements>()),+])
+        )
+    };
+    ($name:ty) => {
+        Type::new(TypeId::of::<$name>(), stringify!($name), TypeContents::PLAIN(stringify!($name)))
+    };
+}
+/// Builds a vec of [`Type`] from a compact invocation, dispatching to the appropriate flavor of [`t!`].
+macro_rules! type_vec {
+    ($name:ident, <$($args:ty),*>) => { vec![$(t!($name<$args>)),*] };
+    ([$($elements:ty),*]) => { vec![$(t!([$elements])),*] };
+    ([$($elements:ty),*]; $len:expr) => { vec![$(t!([$elements; $len])),*] };
+    (($($elements:ty),*)) => { vec![$(t!(($elements, $elements))),*] };
+    ($($names:ty),*) => { vec![$(t!($names)),*] };
+}
+
+lazy_static! {
+    /// The set of registered types. We don't need everything here, just the ones that will be looked up by descriptor
+    /// (i.e., the ones that appear in FFI function generic args).
+    static ref TYPES: Vec<Type> = {
+        let types: Vec<Type> = vec![
+            vec![t!(())],
+            type_vec![bool, char, u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, f32, f64, String],
+            type_vec![(bool, char, u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, f32, f64, String)],
+            type_vec![[bool, char, u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, f32, f64, String]; 1], // Arrays are here just for unit tests, unlikely we'll use them.
```

Comment:
I was going to say that this seems too niche to be useful. 2-tuples of homogeneous type? But the f32 and f64 variants of tuples are actually used for `(eps, del)` pairs.

Not important, but as I'm thinking through this-- since HammingDistance is zero-sized, can I use object_new to create an FfiObject containing a HammingDistance, by passing a slice of length 1 full of garbage? I don't know of a zero-sized type in C, so on the python side I'd end up constructing a non-empty buffer to pass into rust. But then on the rust side, you cast the pointer and it only reads 0 bytes of data to build the struct.

Reply:
Yeah the tuple type generation is a bit lame. I did it this way just to have some examples, and to cover the case of (epsilon, delta) as you mention. (But now I know how to form the product thanks to your last change! 🤣)

I _think_ you could use that trick to construct HammingDistance. You would just pass a NULL for the ptr. We'd need to update the dispatch in `opendp_data__object_new` to include those types. Have we gotten to the point where we need instances of Metrics in client code? We've been getting by with just using the types, but maybe that's breaking down?

Reply:
We don't *really* need metric constructors but it would help partially unjank the identity function. Not particularly in a rush to fix that. We'd also need an accessor for FfiOperation -> FfiObject of domain.